File: aipipe.go
package main

import (
	"context"
	"flag"
	"fmt"
	"io"
	"os"
	"regexp"

	openai "github.com/sashabaranov/go-openai"
)

var (
	GROQ_API_KEY  = os.Getenv("GROQ_API_KEY")
	GROQ_ENDPOINT = os.Getenv("GROQ_ENDPOINT")
	GROQ_MODEL    = os.Getenv("GROQ_MODEL")

	systemMessage = "You are a helpful assistant. If the user has asked for something written, put it in a code block (```), otherwise just provide the answer. If you do use a codeblock, all other text is ignored."
)

func getGroqCompletion(client *openai.Client, userMessage, groqModel string) (string, error) {
	req := openai.ChatCompletionRequest{
		Model: groqModel,
		Messages: []openai.ChatCompletionMessage{
			{
				Role:    openai.ChatMessageRoleSystem,
				Content: systemMessage,
			},
			{
				Role:    openai.ChatMessageRoleUser,
				Content: userMessage,
			},
		},
	}
	resp, err := client.CreateChatCompletion(context.Background(), req)
	if err != nil {
		return "", err
	}
	return resp.Choices[0].Message.Content, nil
}

func extractCodeBlock(completion string) string {
	println("Extracting code block")
	codeBlockPattern := "```[a-zA-Z0-9.]*\\n([\\s\\S]+?)\\n```"
	re := regexp.MustCompile(codeBlockPattern)
	match := re.FindStringSubmatch(completion)
	if match != nil {
		return match[1]
	}
	return completion
}

func isPipedInput() bool {
	fi, err := os.Stdin.Stat()
	if err != nil {
		panic(err)
	}
	return fi.Mode()&os.ModeNamedPipe != 0
}

func main() {
	var codeBlock = flag.Bool("cb", false, "Return only the code block in the completion.")
	flag.Parse()

	flag.Usage = func() {
		fmt.Println("Usage: go run aipipe.go \"query\" > output.txt")
	}

	var prompt string
	if isPipedInput() {
		input, err := io.ReadAll(os.Stdin)
		if err != nil {
			fmt.Fprintln(os.Stderr, "Error reading stdin:", err)
			os.Exit(1)
		}
		prompt = string(input)
	} else {
		args := flag.Args()
		if len(args) < 1 {
			flag.Usage()
			return
		}
		prompt = args[0]
	}

	var client *openai.Client

	config := openai.DefaultConfig(GROQ_API_KEY)
	config.BaseURL = GROQ_ENDPOINT
	client = openai.NewClientWithConfig(config)

	var completion string
	var err error
	completion, err = getGroqCompletion(client, prompt, GROQ_MODEL)

	if err != nil {
		fmt.Fprintln(os.Stderr, "Error:", err)
		os.Exit(1)
	}

	var output string
	if *codeBlock {
		output = extractCodeBlock(completion)
	} else {
		output = completion
	}

	fmt.Print(output)
}

File: aipipe.py
import os
import argparse
import openai
import anthropic
import re
import sys

GROQ_API_KEY = os.getenv('GROQ_API_KEY')
GROQ_ENDPOINT = os.getenv('GROQ_ENDPOINT')
GROQ_MODEL = os.getenv('GROQ_MODEL')

ANTHROPIC_API_KEY = os.getenv('ANTHROPIC_API_KEY')
ANTHROPIC_MODEL = 'claude-3-haiku-20240307'

systemMessage = 'You are a helpful assistant. If the user merely asked a question, do not use a code block. If the user has asked for something written, put it in a code block (```).'

def getGroqCompletion(userMessage, groqModel):
    client = openai.Client(base_url=GROQ_ENDPOINT, api_key=GROQ_API_KEY)
    response = client.chat.completions.create(
        model=groqModel,
        max_tokens=4000,
        messages=[{'role': 'system', 'content': systemMessage},
                  {'role': 'user', 'content': userMessage}]
    )
    return response.choices[0].message.content

def getGpt4Completion(userMessage):
    client = openai.Client()
    response = client.chat.completions.create(
        model="gpt-4-0125-preview",
        messages=[{'role': 'system', 'content': systemMessage},
                  {'role': 'user', 'content': userMessage}]
    )
    return response.choices[0].message.content

def getAnthropicCompletion(userMessage):
    client = anthropic.Anthropic(api_key=ANTHROPIC_API_KEY)
    response = client.messages.create(
        model=ANTHROPIC_MODEL,
        system=systemMessage,
        max_tokens=4000,
        messages=[{'role': 'user', 'content': userMessage}]     
    )
    return response.content[0].text

def extract_code_block(completion):
    code_block_pattern = r'```[a-zA-Z0-9.]*\n([\s\S]+?)\n```'
    match = re.search(code_block_pattern, completion)
    return match.group(1) if match else completion

def main():
    if len(sys.argv) < 2:
        print('Usage: python aifile.py "query" > output.txt')
        return

    parser = argparse.ArgumentParser(description="Get completions from different models.")
    parser.add_argument("prompt", help="The prompt to generate the completion for.")
    parser.add_argument("--codeblock", "--cb", action="store_true", help="Return only the code block in the completion.")
    parser.add_argument("--haiku", action="store_true", help="Use Anthropic Claude model.")
    parser.add_argument("--mx", action="store_true", help="Use Mixtral 8x7b-32768 model.")
    parser.add_argument("--l370", action="store_true", help="Use llama 3 70b.")
    parser.add_argument("--gpt4", action="store_true", help="Use GPT-4 model.")

    args = parser.parse_args()

    useHaiku = args.haiku
    useMixtral = args.mx
    useLlama = args.l370
    useGpt4 = args.gpt4
    codeBlock = args.codeblock
    isatty = sys.stdin.isatty()

    if not isatty:
        filePrompt = sys.stdin.read()

    if args.prompt:
        if(isatty):
            prompt = args.prompt
        else:
            prompt = filePrompt + "\n----\n" + args.prompt
    else:
        prompt = filePrompt

    if not prompt or prompt == '' or prompt.isspace():
        print('Error: No prompt provided', file=sys.stderr)
        return 1

    completion = useHaiku and getAnthropicCompletion(prompt) \
                 or useGpt4 and getGpt4Completion(prompt) \
                 or getGroqCompletion(prompt, useMixtral and 'mixtral-8x7b-32768' or useLlama and 'llama3-70b-8192' or GROQ_MODEL)
    if codeBlock:
        output = extract_code_block(completion)
    else:
        output = completion

    sys.stdout.write(output)

if __name__ == '__main__':
    main()
